retworkx-core/src/shortest_path/bellman_ford.rs

```diff
@@ -0,0 +1,392 @@
+/// ```rust
+/// use retworkx_core::petgraph::Graph;
+/// use retworkx_core::petgraph::prelude::*;
+/// use retworkx_core::dictmap::DictMap;
+/// use retworkx_core::shortest_path::negative_cycle_finder;
+/// use retworkx_core::Result;
+///
+/// let mut graph : Graph<(),i32,Directed>= Graph::new();
+/// let a = graph.add_node(()); // node with no weight
+/// let b = graph.add_node(());
+/// let c = graph.add_node(());
+/// let d = graph.add_node(());
+/// let e = graph.add_node(());
+/// let f = graph.add_node(());
+/// let g = graph.add_node(());
+/// let h = graph.add_node(());
+/// // z will be in another connected component
+/// let z = graph.add_node(());
+///
+/// graph.extend_with_edges(&[
+///     (a, b, 1),
+///     (b, c, 1),
+///     (c, d, 1),
+///     (d, a, 1),
+///     (e, f, 1),
+///     (b, e, 1),
+///     (f, g, -4),
+///     (g, h, 1),
+///     (h, e, 1)
+/// ]);
+/// // a ----> b ----> e ----> f
+/// // ^       |       ^       |
+/// // |       v       |       v
+/// // d <---- c       h <---- g
+///
+/// let res: Result<Option<Vec<NodeIndex>>> = negative_cycle_finder(
+///     &graph, |x| Ok(*x.weight())
+/// );
+/// assert_eq!(res.unwrap().unwrap().len() - 1, 4);
+/// // the first/last node in the cycle appears twice
+/// ```
+pub fn negative_cycle_finder<G, F, K, E>(
+    graph: G,
+    edge_cost: F,
+) -> Result<Option<Vec<G::NodeId>>, E>
+where
+    G: IntoEdges + Visitable + NodeIndexable + NodeCount + IntoNodeIdentifiers,
+    G::NodeId: Eq + Hash + IndexType,
+    F: FnMut(G::EdgeRef) -> Result<K, E>,
+    K: Measure + Copy,
+{
+    let starts: Vec<G::NodeId> = graph.node_identifiers().collect();
+    let res: BellmanFordData<G, Vec<Option<K>>> = inner_bellman_ford(graph, starts, edge_cost)?;
+
+    let BellmanFordData {
+        negative_cycle,
+        scores: _,
+        predecessor,
+    } = res;
+
+    // There is no cycle in the graph
+    if !negative_cycle {
+        return Ok(None);
+    }
+
+    Ok(Some(recover_negative_cycle_from_predecessors(
+        graph,
+        predecessor,
+    )))
+}
+
+fn inner_bellman_ford<G, I, F, K, E, S>(
+    graph: G,
+    starts: I,
+    mut edge_cost: F,
+) -> Result<BellmanFordData<G, S>, E>
+where
+    G: IntoEdges + Visitable + NodeIndexable + NodeCount + IntoNodeIdentifiers,
+    G::NodeId: Eq + Hash + IndexType,
+    F: FnMut(G::EdgeRef) -> Result<K, E>,
+    K: Measure + Copy,
+    S: DistanceMap<G::NodeId, K>,
+    I: IntoIterator<Item = G::NodeId>,
+{
+    let node_count = graph.node_count();
+    let mut in_queue = FixedBitSet::with_capacity(graph.node_bound());
+    let mut scores: S = S::build(graph.node_bound());
+    let mut predecessor: Vec<Option<G::NodeId>> = vec![None; graph.node_bound()];
+    let mut visit_next = VecDeque::with_capacity(graph.node_bound());
+    let zero_score = K::default();
+    let mut relaxation_count: usize = 0;
+
+    for start in starts {
+        scores.put_item(start, zero_score);
+        visit_next.push_back(start);
```

Comment:
Since we're pushing nodes into the queue we should also set `in_queue` to `true`, right?
```suggestion
        visit_next.push_back(start);
        in_queue.set(graph.to_index(start), true);
```